src/os/README.md

```diff
@@ -104,3 +105,125 @@ inspiration for several decisions in this library.
+setters does not matter. For example, given the following short-lived C pointer and
+the related utility functions:
+
+``` c
+// The short-lived C pointer
+typedef void *posix_spawnattr_t;
+
+// The constructor
+int posix_spawnattr_init(posix_spawnattr_t *attr);
+
+// The setters
+int posix_spawnattr_setpgroup(posix_spawnattr_t *attr, pid_t pgroup);
+
+#define	POSIX_SPAWN_RESETIDS 0x0001
+#define	POSIX_SPAWN_SETPGROUP 0x0002
+int posix_spawnattr_setflags(posix_spawnattr_t *attr, short flags);
+
+// The destructor
+int posix_spawnattr_destroy(posix_spawnattr_t *attr);
+```
+
+The corresponding Hack definition should be:
+
+``` hack
+newtype pid_t = int;
+type PosixSpawnFlags = ...;
+type posix_spawnattr_t = shape(
+  'posix_spawnattr_setpgroup' => pid_t,
+  'posix_spawnattr_setflags' => PosixSpawnFlags,
+);
+```
+
+### Ordered repeatable setters
+
+An short-lived C pointers should map to a `vec` definition if the setters can be
+invoked more than once and the order to invoke setters matters. For example,
+given the following short-lived C pointer and the related utility functions:
+
+``` c
+// The short-lived C pointer
+typedef void *posix_spawn_file_actions_t;
+
+// The constructor
+int posix_spawn_file_actions_init(posix_spawn_file_actions_t *file_actions);
+
+// The setters
+int posix_spawn_file_actions_addchdir(posix_spawn_file_actions_t *file_actions, const char *restrict path);
+int posix_spawn_file_actions_adddup2(posix_spawn_file_actions_t *file_actions, int filedes, int newfiledes);
+
+// The destructor
+int posix_spawn_file_actions_destroy(posix_spawn_file_actions_t *file_actions);
+```
+
+The corresponding Hack definition should be:
+
+``` hack
+newtype pid_t = int;
+enum PosixSpawnFlag : int {
+  POSIX_SPAWN_RESETIDS = 0x0001;
+  POSIX_SPAWN_SETPGROUP = 0x0002;
+}
+type posix_spawn_file_actions_t = vec[
+  shape(
+    // Should set one and only one of the following fields
+    ?'posix_spawn_file_actions_addchdir' => string,
+    ?'posix_spawn_file_actions_adddup2' => shape(
+      'filedes' => FileDescriptor,
+      'newfiledes' => int,
+    ),
+  )
+];
+```
+
+Since the original C function `posix_spawn_file_actions_adddup2` accepts three
+parameters, the second parameter and the third parameter are packed
+into a `shape`.
+
+Note even though all the setters are defined in the same `shape`, the user
+should set one and only one of the fields. A runtime check should be performed
+by the HSL implementation, ensuring only one of them is set.
+
+The above `shape` encoding should be
+assumed as the following union type.
+
+``` hack
+// If union types is enabled
+type posix_spawn_file_actions_t = vec[
+  shape(
+    'posix_spawn_file_actions_addchdir' => string,
+  )|
+  shape(
+    'posix_spawn_file_actions_adddup2' => shape(
+      'filedes' => FileDescriptor,
+      'newfiledes' => int,
+    ),
+  ),
+];
+```
+
+However we cannot actually encode the setters as a true union type because union types are not
+enabled for the HSL code base.
```

Comment:
We should design for the language we have/reasonably expect to have; if the end result of following the rules does not work in the language, we need to revisit the rules.

Reply:
OK, I created #174, which includes the encoding of the setters as an interface to eliminate the need to union types.